pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java

```diff
@@ -300,7 +329,33 @@ public void invalidateAll() {
      */
     protected void execute(Runnable task, CompletableFuture<?> future) {
         try {
-            executor.execute(task);
+            // Wrap the original task, so we can record the thread on which it is running
+            TaskWrapper taskWrapper = new TaskWrapper(task);
+            executorWatchDog.execute(() -> {
+                // Delegate the running of the task to the executor
+                Future<?> f = executor.submit(taskWrapper);
+                long start = System.currentTimeMillis();
+                while (!f.isDone()) {
+                    try {
+                        f.get(5000, TimeUnit.MILLISECONDS);
+                    } catch (Exception e) {
+                        long now = System.currentTimeMillis();
+                        Thread taskThread = taskWrapper.getTaskThread();
+                        if (taskThread != null) {
+                            final StackTraceElement[] stackTrace = taskThread.getStackTrace();
+                            if (stackTrace.length > 0) {
+                                StringBuilder stackTraceString = new StringBuilder();
+                                for (StackTraceElement elem : stackTrace) {
+                                    stackTraceString.append(elem);
+                                    stackTraceString.append("\n");
+                                }
+                                log.info("Metadata store task took {} milli secs so far. Stack trace: {}",
```

Comment:
It looks like custom debug snippet, not sure if it's good to add to mainline.

Reply:
@MarvinCai the metadata-store callback thread uses a single-threaded executor (https://github.com/apache/pulsar/blob/7058d771b4403f7e6d0036cc71691c562c745131/pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java#L79). If a blocking call is made to metadata-store API, that will result in a deadlock.

It is non-trivial to chase down such deadlock issues in production environments. It is also not trivial to catch such cases during code review. The log statement is not a custom debug snippet, but it was added deliberately to monitor the callback tasks and log if a given task is taking too long, along with the stack trace. This helps identify which code path is causing the deadlock and the code can then be corrected easily.